### DISCUSSION TITLE
운영체제 - CPU 스케쥴링 - 진호 && 호제

### DISCUSSION THREAD

**Discussion by zini9188** [Shortudy-10th/tech-for-developer#120] [질의 응답]
* CPU 스케쥴러는 무엇인가요?
* CPU 스케줄링의 성능 척도에는 어떤 것들이 있나요?
* 운영체제가 스케줄링 결정을 내리는 시점과 방식에 대해 설명해주세요.
* 스케줄링의 목적과 저, 중, 고수준 단계에 대해 말해주세요.
* 스케줄러가 어떤 프로세스에 우선적으로 CPU를 할당할지 스케줄링 시 고려하는 사항에 대해 말해주세요.

**Comment by zini9188** [comment]
❓ 스케줄링의 목적과 저, 중, 고수준 단계에 대해 말해주세요.

💬 답변
### 스케줄링의 목적
- 모든 프로세스가 공평하게 작업하도록 하여 안정적으로 작동하기 위해 존재
- 특정 프로세스가 시스템 자원을 독점하거나 파괴하는 것을 막기 위해 중요도에 따라 우선순위를 배정

> 공평성
- 모든 프로세스가 자원을 공평하게 배정하며 배제되는 프로세스가 없어야 함
> 효율성
- 시스템 자원이 유휴 시간이 없이 사용되도록 스케줄링
- 유휴 자원을 사용하려는 프로세스에 우선권 제공
> 안정성
- 시스템 자원을 점유하거나 파괴하려는 프로세스로부터 자원을 보호
> 확정성
- 프로세스가 증가해도 시스템이 안정적으로 작동하도록 조치해야 함
> 반응 시간 보정
- 시스템은 적절한 시간 안에 프로세스의 요구에 반응해야 함
> 무한 연기 방지
- 특정 프로세스 작업이 무한히 연기되면 안됨

![img1 daumcdn](https://github.com/Shortudy-10th/tech-for-developer/assets/83208807/53de25bf-f322-488b-8a66-bc5480541a24)

### 고수준 스케줄링
- 가장 큰 틀에서 이루어지는 스케줄링으로 `작업 스케줄링`, `장기 스케줄링` 이라고도 함
- 시스템 내의 전체 작업 수를 조절
  - 작업의 의미는 운영체제에서 다루는 일의 가장 큰 단위로, 1개 또는 여러 개의 프로세스로 구성

### 중간 수준 스케줄링
- 시스템의 부하를 조절하기 위한 스케줄링
- 시스템의 과부하가 걸려 프로세스 수를 조절해야 한다면, 이미 활성화된 프로세스 중 일부를 보류 상태로 보내고 여유가 생기면 복구

### 저수준 스케줄링
- 가장 작은 단위의 스케줄링으로 `단기 스케줄링`이라고도 함
- 실제 작업이 이루어짐

**Comment by Jihn0118** [comment]
❓ 스케줄러가 어떤 프로세스에 우선적으로 CPU를 할당할지 스케줄링 시 고려하는 사항에 대해 말해주세요.

💬 답변

**스케줄링 정책, 프로세스의 우선순위, 프로세스의 상태, I/O 작업 및 Blocking, switching 오버헤드의 최소화** 등이 있습니다.

이러한 요소들을 종합적으로 고려하여 스케줄러는 CPU 할당을 결정하고, 시스템 성능을 최적화하려고 합니다. 
다양한 스케줄링 알고리즘과 정책이 개발되어 있으며, 시스템의 목표와 특성에 따라 적합한 스케줄링 방식을 선택하는 것이 중요합니다.

**Reply by zini9188** [reply]
❓ CPU 스케줄러에서 오버헤드 최소화의 중요성과 방법에 대해 설명해주세요

**Reply by zini9188** [reply]
❓ I/O 작업과 Blocking이 스케줄링에 미치는 영향에 대해 설명해주세요

**Reply by zini9188** [reply]
❓ 프로세스의 상태 변화에 따른 스케줄링 영향에 대해 설명해주세요

**Reply by Jihn0118** [reply]
1. 오버헤드 최소화의 중요성과 방법
오버헤드는 프로세스 간 전환 등으로 인해 CPU 시간이 낭비되는 상황을 의미하는데, 이를 최소화하여 시스템 성능을 개선할 수 있다.
오버헤드를 최소화하는 방법은 시스템의 목표와 특성에 따라 적합한 스케줄링 알고리즘을 선택해야 합니다.

2. I/O 작업과 Blocking이 스케줄링에 미치는 영향
대기 상태에 있는 프로세스가 I/O 작업을 수행 중인 경우, 스케줄러는 해당 프로세스를 다른 프로세스로 교체할 수 있다. 이렇게 함으로써 CPU가 유휴 상태에 머무르지 않고 다른 작업을 수행할 수 있다.

3. 프로세스의 상태 변화에 따른 스케줄링 영향
프로세스는 준비 상태(Ready), 실행 상태(Running), 대기 상태(Waiting) 등 다양한 상태를 가집니다. 스케줄러는 현재 상태에 따라 어떤 프로세스에 CPU를 할당할지 결정한다. 준비 상태에 있는 프로세스 중에서 어떤 프로세스를 선택할 것인지 고려한다.

**Comment by zini9188** [comment]
❓ CPU 스케줄링의 성능 척도에는 어떤 것들이 있나요?

💬 답변
CPU 스케줄링의 성능 척도는 크게 시스템 관점과 사용자 관점으로 나뉜다.

### 시스템 관점
성능 척도  | 설명
--    | --
CPU 이용률 | CPU가 쉬지 않고 일한 시간의 비율. 높을 수록 좋다. CPU 스케줄링에 따라 이용률이 변경될 수 있음
처리량 | 단위시간당 처리량. 시스템 입장에서 얼마나 많은 일을 했는가를 나타내며 높을수록 좋다

### 사용자 관점
성능 척도 | 설명
-- | --
소요시간, 반환시간 | CPU burst에 들어와서, I/O burst가 될 때까지의 전체 시간의 합. Ready Queue에서 기다린 시간 + 실제로 CPU를 사용한 시간
대기시간 | CPU를 사용하기 위해 기다린 전체 시간의 합. CPU burst에 들어와서 프로세스는 CPU를 얻고 뺏기를 반복하는 데, I/O burst 상태에 되기 직전까지 Ready Queue에서 CPU 사용을 위해 기다린 전체 시간의 합
응답 시간 | 특정 프로세스가 CPU를 사용하기 위해 들어와서 CPU를 최초로 사용하기까지 걸리는 시간. CPU burst로 들어와 최초로 CPU를 얻기까지 걸리는 시간. 대부분의 프로세스는 preemptive이므로 이 시간이 응답 시간으로 측정된다.

**Comment by zini9188** [comment]
❓ CPU 스케줄러는 무엇인가요?

💬 답변
- 운영 체제의 일부로, 여러 프로세스들 중 어떤 프로세스가 CPU를 사용하여 실행될 것인지 결정하는 역할을 담당하는 소프트웨어
- 프로세스의 우선순위, 실행 시간, 대기 시간 등과 같은 다양한 요소를 고려하여 프로세스 수행을 스케줄
- 장기, 단기, 중간 스케줄러의 유형으로 나뉨
- 성능 척도 및 사용자 요구를 만족시키기 위해 FCFS, SJF, 우선순위, 라운드 로빈 등의 알고리즘이 사용

**Comment by Jihn0118** [comment]
❓ 운영체제가 스케줄링 결정을 내리는 시점과 방식에 대해 설명해주세요.

💬 답변

### 1. 스케줄링 결정을 내리는 시점
- **프로세스 생성**: 새로운 프로세스가 생성되거나 실행 가능한 상태로 변경될 때, 
   운영체제는 이 프로세스에 CPU를 할당할지 여부를 결정합니다.

- **프로세스 종료**: 프로세스가 종료되면 해당 프로세스에 할당된 CPU 자원을 반환하고, 
   다음으로 실행될 프로세스를 선택합니다.

- **인터럽트 및 시그널**: 외부 이벤트나 프로세스 내부에서 발생한 인터럽트 또는 시그널에 응답하여, 운영체제는 현재 실행 중인 프로세스를 일시 중단하고 다른 프로세스로 전환할 수 있습니다.

- **시간 할당량 초과**: CPU 할당 시간이 만료된 경우, 해당 프로세스는 준비 상태로 이동하고 다른 프로세스에게 CPU를 할당합니다. 
  이러한 방식은 프로세스 간의 공정성을 유지하기 위해 사용됩니다.

- **I/O 요청**: 프로세스가 I/O 작업을 요청하면, 해당 프로세스는 대기 상태로 전환되고, 다른 실행 가능한 프로세스가 CPU를 할당 받을 수 있습니다. 
  I/O 작업이 완료되면 해당 프로세스는 준비 상태로 이동하여 다시 CPU를 할당 받을 수 있습니다.

- **우선순위 변경**: 프로세스의 우선순위가 변경되거나, 운영체제가 더 높은 우선순위의 프로세스에게 CPU를 할당하도록 결정할 수 있습니다.


### 2. 스케줄링 방식
- **선점형 스케줄링**
  -  프로세스가 실행 중에 다른 프로세스가 CPU를 강제로 빼앗아 사용할 수 있는 스케줄링 방식
  -  시간 할당량이나 우선순위에 따라 프로세스를 전환
  -  Round Robin, Priority Scheduling 등이 선점형 스케줄링의 예

- **비선점형 스케줄링**
  -  프로세스가 CPU를 스스로 반납하기 전까지 실행을 계속할 수 있는 스케줄링 방식
  -  프로세스가 끝나거나 I/O 요청을 하면 CPU를 반환
  -  FCFS (First-Come, First-Served), SJF (Shortest Job First) 등이 비선점형 스케줄링의 예
